### PR TITLE
Use libuv for tempdir function

### PR DESCRIPTION
**Pull request by musm**
Use libuv's implementation for `tempdir` for both linux and windows. This fixes the inconsistency between the windows and linux version, where the windows version does not currently include a trailing slash when calling `tempdir` compared to the linux implementation.

### CODE REVIEW

**Comment by musm** [comment]
Is this something we care about?

**Comment by StefanKarpinski** [comment]
Yes, but too late for 1.2.

**Review by stevengj — APPROVED** [review]

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: This would fail if `temppath` were the root directory, e.g. `C:\`, but I guess that shouldn't be possible.

**Comment by musm** [comment]
came up here https://discourse.julialang.org/t/tempdir-differences-in-windows-and-linux/24930

File: base/file.jl
Comment (musm) [reply base/file.jl]: Good point, this would occur if `TMP` env variable was set to `C:\\`

Comment (musm) [reply base/file.jl]: Latest commit checks that the folder is not a root directory and if it is a root directory does not strip the trailing slash

Comment (musm) [reply base/file.jl]: This fixes the case where `TMP` is set to `C:\\`  (kinda pathological, but in theory possible)

File: base/file.jl
Comment (musm) [inline base/file.jl]: I can also write this is as
```julia
    isroot = Bool(ccall((:PathIsRootW,:Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
    !isroot && temppath[lentemppath] == 0x005c ?  resize!(temppath,lentemppath-1) :  resize!(temppath,lentemppath)
```

Comment (stevengj) [reply base/file.jl]: Or
```jl
isroot = Bool(ccall((:PathIsRootW,:Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
resize!(temppath, !isroot && temppath[lentemppath] == 0x005c ? lentemppath-1 : lentemppath)
```

File: base/file.jl
Comment (vtjnash) [inline base/file.jl]: `Shlwapi` is a very large dependency that we don't appear to use anywhere else. It would be nice avoid instantiating it just for this. There should already be a regex for this in path.jl for splitdrive.

File: base/file.jl
Comment (vtjnash) [reply base/file.jl]: Or
```julia
isroot = Bool(ccall((:PathIsRootW, :Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
(!isroot && temppath[lentemppath] == 0x005c) && (lentemppath -= 1)
resize!(temppath, lentemppath)
```

File: test/file.jl
Comment (vtjnash) [inline test/file.jl]: excess space?

File: base/file.jl
Comment (musm) [reply base/file.jl]: Sure, is there any overhead to instatitating it or is the preference to keep as much of the code in Julia? I don't see any real overhead to calling the library since it's loaded by default.

Comment (musm) [reply base/file.jl]: Updated code with a julia version of `isroot`. Let me know if this is preffered

File: base/file.jl
Comment (musm) [inline base/file.jl]: Should I move this to `path.jl` ?

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: Isn't the common case for `temppath` to end with `\`, which means that this makes an extra copy of the string in the common case?

Comment (musm) [reply base/file.jl]: Yea that is the common case. I'm not sure how'd go about avoiding the extra copy when using the julia  `_isroot` implementation.

Comment (stevengj) [reply base/file.jl]: We could use a modified isroot regex that is missing the trailing slash, and test it on the string after we strip the slash—if it matches, add the slash back.

File: base/file.jl
Comment (musm) [inline base/file.jl]: Should I move this to `path.jl`?

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: Shouldn't it check that there is a trailing `\` first?  e.g.
```jl
resize!(temppath,lentemppath - (temppath[lentemppath]==0x005c))
```

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: If `m.captures[2] == "\\"` then we don't need to add another `\`, right?    So we actually just want `isempty(m.captures[2])` for this case.    You also don't need the `... ? true : false`, since you can just return `...`.

Maybe it's better to just inline
```jl
# check whether path is the drive root or UNC path, missing the trailing backslash:
isroot = isempty(match(r"^([^\\]+:|\\\\[^\\]+\\[^\\]+|\\\\\?\\UNC\\[^\\]+\\[^\\]+|\\\\\?\\[^\\]+:|)(.*)$", path).captures[2])
```

File: base/file.jl
Comment (musm) [reply base/file.jl]: According to their docs https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-gettemppathw
the returned string ends with a backslash.

Comment (stevengj) [reply base/file.jl]: I like how their docs say that it ends with a backslash and immediately contradicts itself with the example: ``` The returned string ends with a backslash, for example, "C:\TEMP". ```

File: base/file.jl
Comment (musm) [reply base/file.jl]: if `path = "C:"` then   `m.captures[2] = ""`   if `path = "C:\\"` then  `m.captures[2]== "\\"` . Path paths are roots.      Because `GetTempPathW` always adds a backslash `isempty(m.captures[2])`  is not needed, however I added it since the function tests if the path is rooted.

File: base/file.jl
Comment (musm) [reply base/file.jl]: I know, someone at MSFT needs to fix that....    (I tested several combinations and the function always adds backslash...)

File: base/file.jl
Comment (musm) [reply base/file.jl]: I'm not sure how useful this function is in general, but it could be made to check is the path is a root for   non -windows OS,   by checking if the path starts with `/` ?

**Comment by musm** [comment]
@stevengj I addressed your feedback, let me know if these changes now have your approval. Thanks for the review.


I also think GetTempPathW could probably just be implemented in Julia, without too much difficulty. Probably best for another PR, however.

> 
> The GetTempPath function checks for the existence of environment variables in the following order and uses the first path found:
> 
> The path specified by the TMP environment variable.
> The path specified by the TEMP environment variable.
> The path specified by the USERPROFILE environment variable.
> The Windows directory.
> Note that the function does not verify that the path exists, nor does it test to see if the current process has any kind of access rights to the path. The GetTempPath function returns the properly formatted string that specifies the fully qualified path based on the environment variable search order as previously specified. The application should verify the existence of the path and adequate access rights to the path prior to any use for file I/O operations.
> Symbolic link behavior—If the path points to a symbolic link, the temp path name maintains any symbolic links.
>

**Comment by vtjnash** [comment]
If we want to go that route, we could switch to just using `uv_os_tmpdir` everywhere (it already handle this case).

**Comment by musm** [comment]
Any reason not to go the libuv route? Seems simpler, then custom casing Windows.

**Comment by musm** [comment]
I was testing `uv_os_tmpdir`

```julia
function _tempdir()
    path_max = 1024
    buf = zeros(UInt8, path_max)
    sz = Ref{Csize_t}(path_max)
    rc = ccall(:uv_os_tmpdir, Cint, (Ptr{UInt8}, Ptr{Csize_t}), buf, sz)
    if rc == 0
        @show buf, sz[]
        resize!(buf, sz[])
        return String(buf)
    else
        error("unable to retrieve temporary directory directory")
    end
end
  
```

it seems to work fine EXCEPT when `ENV["TMP"]= ""` ,  in which case garbage is spit out. `GetTempPathW ` would spit out the correct value.  I think there might be a bug in  `uv_os_tmpdir` ?

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: StringVector?

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: Assertion or systemerror?

File: base/file.jl
Comment (musm) [reply base/file.jl]: sure

File: base/file.jl
Comment (musm) [reply base/file.jl]: I'm not 100% sure which is best.  There are two scenarios were `rc!=0`: Invalid inputs to `uv_os_tmpdir`, for which `@assert` is appropriate since this should never happen; and  possible errors  related to buffer having e.g. too small a length, which shouldn't happen, but might, for which throwing an error is probably best?

From the docs

```
 On success size is set to the string length of buffer (which does not include the terminating null). On UV_ENOBUFS failure size is set to the required length for buffer, including the null byte.
```

**Comment by musm** [comment]
According to https://github.com/JuliaLang/julia/commit/84e68f3a7ef7d888af8a580dc52f27ff9209f4ef by @Keno  it seems like we should be using RefValue instead of Ref here?  (there are to instances where Ref is used)

File: base/file.jl
Comment (musm) [reply base/file.jl]: Apparently already reported https://github.com/MicrosoftDocs/feedback/issues/898

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: Use e.g. `uv_error("tempdir", rc)`, which checks the error code and gives an appropriate message.

**Comment by musm** [comment]
Ok I updated the code and this is now ready for a final review. Last commit handles  the UV_ENOBUFS case with tests. I think this is now good to merge.

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: needs to be length `path_max+1` since libuv copies the trailing NUL char: https://github.com/JuliaLang/libuv/blob/26dbe5672c33fc885462c509fe2a9b36f35866fd/src/unix/core.c#L1105-L1106

File: base/file.jl
Comment (stevengj) [inline base/file.jl]: No `- 1`, similar to above.

File: base/file.jl
Comment (stevengj) [reply base/file.jl]: (either that or set `sz` to `path_max` … it's okay if its too small since we reallocate in the unlikely event of such a long path.)

Comment (musm) [reply base/file.jl]: Thanks.
I removed the `sz` initialization, since it's not needed.

**Review by stevengj — APPROVED** [review]

**Comment by musm** [comment]
unrelated error in testset Distributed on macos

File: test/file.jl
Comment (musm) [inline test/file.jl]: For reference I'm pretty sure there's a bug in the libuv function that hits an edge case that shouldn't happen, which is why the test here is for 251 chars and does not test longer paths.

hitting this branch I believe https://github.com/libuv/libuv/blob/c4e9657d59723a6273ce68bda935680e1b3596c5/src/win/util.c#L1186
Looks like the issue is that the code is using MAXPATH+1 instead of MAXPATH

**Comment by musm** [comment]
Before merging let me push an update to account for the latest changes on master

**Comment by musm** [comment]
Test failures are flukes (distributed fluke and a timeout fluke during distributed). Is this  ok to merge ?

**Comment by musm** [comment]
good to merge?

**Comment by musm** [comment]
Test failures are unrelated, due to downloads

**Comment by musm** [comment]
@vtjnash can you sign off if this looks good to you? 

FYI the `@test_broken` is fixed by my PR to libuv

File: base/file.jl
Comment (vtjnash) [inline base/file.jl]: Change to use `tempdir()` for `d`?

File: base/file.jl
Comment (vtjnash) [inline base/file.jl]: In the `tempname` doc string below, it might be clearer to add, "This can lead to security holes." and recommend opening the file with `O_EXCL`, if this is a concern (in addition to the suggestion to use `mktemp`).

Aside: the windows implementation of `mktemp` appears to be wrong because it fails to heed this warning.

**Review by vtjnash — APPROVED** [review]

File: base/file.jl
Comment (musm) [reply base/file.jl]: Will make these changes in a separate PR.

Comment (vtjnash) [reply base/file.jl]: sgtm. feel free to merge.

**Comment by musm** [comment]
> sgtm. feel free to merge.


Someone with merge powers want to take care of this 😉

**Pull request merged**
